### PR TITLE
binloginfo: fix a data race on global.Config() in the unit test

### PR DESCRIPTION
**Pull request by tiancaiamao**
```
WARNING: DATA RACE
Read at 0x000002591538 by goroutine 20:
  github.com/pingcap/tidb/sessionctx/binloginfo.(*BinlogInfo).WriteBinlog()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/sessionctx/binloginfo/binloginfo.go:118 +0x667
  github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).writeFinishBinlog.func1()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:683 +0x6b

Previous write at 0x000002591538 by goroutine 53:
  github.com/pingcap/tidb/sessionctx/binloginfo_test.(*testBinlogSuite).TestIgnoreError()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/sessionctx/binloginfo/binloginfo_test.go:395 +0x499
  github.com/pingcap/tidb/util/testkit.(*TestKit).Exec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:130 +0x133
  github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:159 +0x91
  github.com/pingcap/tidb/sessionctx/binloginfo_test.(*testBinlogSuite).TestIgnoreError()
      /home/jenkins/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/sessionctx/binloginfo/binloginfo_test.go:378 +0x275
  github.com/pingcap/tidb/sessionctx/binloginfo_test.(*testBinlogSuite).TestIgnoreError()
```

I should make a atomic variable, rather than operate on config.Binlog directly.

```
goroutine1:
       	cfg := config.GetGlobalConfig()
	if cfg.Binlog.IgnoreError {

goroutine2 (in test code):

	cfg := config.GetGlobalConfig()
	cfg.Binlog.IgnoreError = true
```

I'll update https://github.com/pingcap/tidb/pull/6546 when this is merged.

@coocood @shenli

### CODE REVIEW

**Comment by coocood** [comment]
ignoreError is a config, erroStopped is a state, why should they share the same flag?

**Comment by tiancaiamao** [comment]
ignoreError is a state... the status machine is
ignoreError -> errorStopped -> ignoreError -> errorStopped ...

If the initial state is not ignoreError, those logic would be ignored totally. @coocood

**Comment by zz-jason** [comment]
@tiancaiamao How about s/ignoreError/errorIgnored/ ?

**Comment by tiancaiamao** [comment]
If the binlog never meet any error,  `ignoreErr` still means "if binlog meets error, ignore it". 
what's problem? @coocood

**Comment by tiancaiamao** [comment]
PTAL @shenli @coocood @zz-jason

File: sessionctx/binloginfo/binloginfo.go [hunk 1]
```diff
@@ -114,11 +119,11 @@ func (info *BinlogInfo) WriteBinlog(clusterID uint64) error {
...
 
 	if err != nil {
-		cfg := config.GetGlobalConfig()
-		if cfg.Binlog.IgnoreError {
+		if atomic.LoadUint32(&ignoreError) == 1 {
 			log.Errorf("critical error, write binlog fail but error ignored: %s", errors.ErrorStack(err))
 			metrics.CriticalErrorCounter.Add(1)
-			atomic.StoreUint32(&errorStopped, 1)
+			// If error happens once, we'll stop writing binlog.
+			atomic.CompareAndSwapUint32(&stopped, valueStopped, valueStopped+1)
```
Comment (coocood) [inline sessionctx/binloginfo/binloginfo.go]: How about just `atomic.StoreUint32(1)`?

File: sessionctx/binloginfo/binloginfo.go [hunk 2]
```diff
@@ -74,17 +73,23 @@ func GetPrewriteValue(ctx sessionctx.Context, createIfNotExists bool) *binlog.Pr
...
+}
+
+func SetIgnoreError() {
+	atomic.StoreUint32(&ignoreError, 1)
 }
 
 // WriteBinlog writes a binlog to Pump.
 func (info *BinlogInfo) WriteBinlog(clusterID uint64) error {
-	if atomic.LoadUint32(&errorStopped) != 0 {
+	valueStopped := atomic.LoadUint32(&stopped)
```
Comment (coocood) [inline sessionctx/binloginfo/binloginfo.go]: How about
```
if atomic.LoadUint32(&stopped) != 0
```

File: sessionctx/binloginfo/binloginfo.go (hunk 1 above)
Comment (tiancaiamao) [reply sessionctx/binloginfo/binloginfo.go]: I write this code deliberately. @coocood 

Consider this scenario:

```
txn1 ... WriteBinlog() ...  check stopped = 0 ... continue
txn2 ... WriteBinlog() ...  check stopped = 0 ... continue
....
txn2 ... run here ... atomic.StoreUint32(&stopped, 1) ... so stopped is 1

http.ResetStopFlag()     // After this, we all think stop flag has been reset... but ...

txn1 ... run here ... atomic.StoreUint32(&stopped, 1) .... 

What the fuck? ResetStopFlag operation is gone,  we lost update!
```

File: sessionctx/binloginfo/binloginfo.go [hunk 3]
```diff
@@ -74,17 +73,25 @@ func GetPrewriteValue(ctx sessionctx.Context, createIfNotExists bool) *binlog.Pr
 	return v
 }
 
-var errorStopped uint32
+var stopped uint32
```
Comment (shenli) [inline sessionctx/binloginfo/binloginfo.go]: Add comments for the two variables.

File: sessionctx/binloginfo/binloginfo.go [hunk 4]
```diff
@@ -74,17 +73,25 @@ func GetPrewriteValue(ctx sessionctx.Context, createIfNotExists bool) *binlog.Pr
...
+var ignoreError uint32
 
-// ResetErrorStopFlag resets the errorStopped flag.
-func ResetErrorStopFlag() {
-	atomic.StoreUint32(&errorStopped, 0)
-	log.Warn("[binloginfo] reset error stop flag")
+// ResetStopFlag resets the errorStopped flag.
+func ResetStopFlag() {
```
Comment (shenli) [inline sessionctx/binloginfo/binloginfo.go]: How about DisableStopFlag?

File: sessionctx/binloginfo/binloginfo.go (hunk 3 above)
Comment (shenli) [reply sessionctx/binloginfo/binloginfo.go]: `stopped` is not a good name. `skipBinlog` is better.

**Comment by shenli** [comment]
LGTM

**Comment by tiancaiamao** [comment]
/run-all-tests
@zz-jason @coocood

**Pull request merged**